File: 02_read.py
# # Reading and Writing Data

# GAD NOTE: 1.
# GAD NOTE: Changed all generated HDFS files to relative path 'practice' instead
# GAD NOTE: of relative path 'duocar'--hopefully that's clearer.
# GAD NOTE: 2.
# GAD NOTE: The local directory data was missing.  Added it back, and changed
# GAD NOTE: the local file to read into pandas to demographics.txt.
# GAD NOTE: 3.
# GAD NOTE: Purging practice files output files after the run.
# GAD NOTE: 4.
# GAD NOTE: You'll see other minor changes inline.

# Spark can read and write data in a variety of formats.

# * Working with text files
# * Working with delimited text files
# * Working with Parquet files
# * Working with pandas DataFrames
# * Generating a Spark DataFrame


# ## Setup

# Create a SparkSession:
from pyspark.sql import SparkSession
spark = SparkSession.builder.appName('read').master('local').getOrCreate()

# **Note:** The default filesystem in CDH (and by extension CDSW) is HDFS.


# ## Working with text files

# The `text` method of the `DataFrameReader` class reads each line of a text file 
# into a row of a DataFrame with a single column named *value*.

riders_txt = spark.read.text('/duocar/raw/riders/')
riders_txt.show(5)
riders_txt.head(5)

# In this case, we would need to apply further transformations to parse each line.

# Create a subdirectory in the user's home directory of HDFS:

!hdfs dfs -rm -r -skipTrash practice

!hdfs dfs -mkdir practice

# The `text` method of the `DataFrameWriter` class writes each row of a DataFrame 
# with a single string column into a line of text file.

riders_txt.write.text('practice/riders_text')

# Note that this write is to a *relative path*, in the user's home directory.

!hdfs dfs -cat practice/riders_text/* | head -n 5

# **Note:** Do not worry about the `cat: Unable to write to output steam.` message.

# The `text` method can also write a compressed file.

riders_txt.write.text('practice/riders_text_compressed', compression='bzip2')

!hdfs dfs -ls practice/riders_text

!hdfs dfs -ls practice/riders_text_compressed

# **Reference:** <https://www.cloudera.com/documentation/enterprise/latest/topics/introduction_compression.html>


# ## Working with delimited text files

# The rider data is actually a comma-delimited text file.  The `csv` method of the 
# `DataFrameWriter` class reads a delimted text file.

spark.read.\
  csv('/duocar/raw/riders/', sep=',', header=True, inferSchema=True).\
  show(5)

# This is a convenience function for the more general syntax:

riders = spark.read.\
  format('csv').\
  option('sep', ',').\
  option('header', True).\
  option('inferSchema', True).\
  load('/duocar/raw/riders/')

# **Note:** If you use either method with `inferSchema` set to `True`, then Spark 
# assumes that a header row occurs in *every* file in the data directory you load.

# You can manually specify the schema instead of inferring it from the header 
# row and column values:

from pyspark.sql.types import *

schema = StructType([
    StructField("id", StringType()),
    StructField("birth_date", DateType()),
    StructField("start_date", DateType()),
    StructField("first_name", StringType()),
    StructField("last_name", StringType()),
    StructField("sex", StringType()),
    StructField("ethnicity", StringType()),
    StructField("student", IntegerType()),
    StructField("home_block", StringType()),
    StructField("home_lat", DoubleType()),
    StructField("home_lon", DoubleType()),
    StructField("work_lat", DoubleType()),
    StructField("work_lon", DoubleType())
])

spark.read.\
  format('csv').\
  option('sep', ',').\
  option('header', True).\
  schema(schema).\
  load('/duocar/raw/riders/').\
  show(5)


# Write the file to a comma-delimited file:

riders.write.csv('practice/riders_csv', sep=',')

!hdfs dfs -ls practice/riders_csv

!hdfs dfs -cat practice/riders_csv/* | head -n 5


# ## Working with Parquet files

riders.write.parquet('practice/riders_parquet')

!hdfs dfs -ls practice/riders_parquet


# ## Working with pandas DataFrames

import pandas as pd

demographics_pd = pd.read_csv('data/demographics.txt', sep='\t')

# Note this read is from the local directory, not from HDFS.

demographics_pd.head()

spark.createDataFrame(demographics_pd).show(5)

# **Developer Note:** This does not work in Python 3.
# Different versions of Python on driver (3.6) and worker (2.6).

# Use the `toPandas` method to read a Spark DataFrame into a pandas DataFrame.

riders_pd = riders.toPandas()

riders_pd.head()

# **WARNING:** Use this with caution as you may use all your available memory!


# ## Generating a Spark DataFrame

# Sometimes we need to generate a Spark DataFrame from scratch, for example, 
# for testing purposes.

# Use the `range` method to generate a sequence of integers and add new 
# columns as appropriate.

spark.range(1000).show(5)

# Use the `rand` function to generate a uniform random variable:

from pyspark.sql.functions import rand

spark.range(1000).\
  withColumn('uniform', rand(12345)).\
  show(5)

# or a Bernoulli random variable with `p = 0.25`:

bern_df = spark.range(1000).\
  withColumn('Bernoulli', (rand(12345) > 0.75).cast("int"))
  
# Summary using functional style:

bern_df.groupby('Bernoulli').count().show()

# Summary using SQL style:

bern_df.createOrReplaceTempView("bern")

spark.sql("SELECT Bernoulli, COUNT(*) AS count\
    FROM bern\
    GROUP BY Bernoulli").\
  show()

# Use the `randn` function to generate a normal random variable:

from pyspark.sql.functions import randn

ran_df = spark.range(1000).\
  withColumn('normal', 42 +  2 * randn(54321))
ran_df.show(5)
ran_df.describe('id', 'normal').show()

# ## Cleanup

# Remove practice directory from HDFS.

!hdfs dfs -rm -r -skipTrash practice/riders_text

# Stop the `SparkSession`.

spark.stop()

File: 02_read.R
# # Reading and Writing Data

# ## Setup

library(sparklyr)
spark <- spark_connect(master = "local", app_name = "read")


# ## Working with delimited text files

# The rider data is a comma-delimited text file
# Use the `spark_read_csv()` function to read it into a Spark DataFrame

# ### Infer the schema

# With only the required arguments:

riders <- spark_read_csv(
  sc = spark,
  name = "riders",
  path = "/duocar/raw/riders/"
)

# Showing some more arguments, which are set to the default values:

riders <- spark_read_csv(
  sc = spark,
  name = "riders",
  path = "/duocar/raw/riders/",
  header = TRUE,
  infer_schema = TRUE,
  delimiter = ","
)

# Show the schema and first few rows of data:
riders

# The returned object (`riders` in this example) is a `tbl_spark`.

class(riders)

# This represents a remote Spark DataFrame. It is not loaded in R's memory.
# Notice that only the first few rows of data are shown.


# ### Manually specify the schema

riders2 <- spark_read_csv(
  sc = spark,
  name = "riders",
  path = "/duocar/raw/riders/",
  infer_schema = FALSE,
  columns = c(
    id = "character",
    birth_date = "character",
    start_date = "character",
    first_name = "character",
    last_name = "character",
    sex = "character",
    ethnicity = "character",
    student = "integer",
    home_block = "character",
    home_lat = "numeric",
    home_lon = "numeric",
    work_lat = "numeric",
    work_lon = "numeric"
  )
)

riders2


# Write the file to a tab-delimited file:

system("hdfs dfs -rm -r -skipTrash practice")

system("hdfs dfs -mkdir practice")

spark_write_csv(
  riders,
  path = "practice/riders_tsv",
  delimiter = "\t"
)

system("hdfs dfs -ls practice/riders_tsv")

system("hdfs dfs -cat practice/riders_tsv/* | head -n 5")

# **Note:** Do not worry about the `cat: Unable to write to output steam.` message.


# ## Working with Parquet files

spark_write_parquet(
  riders,
  path = "practice/riders_parquet"
)

system("hdfs dfs -ls practice/riders_parquet")

# Note that the schema is stored with the data

riders_parquet <- spark_read_parquet(
  sc = spark,
  name = "riders_parquet",
  path = "practice/riders_parquet"
)
riders_parquet


# ## Working with Hive Tables

# Use the `DBI::dbGetQuery()` function to run SQL queries:

library(DBI)

dbGetQuery(spark, "SHOW DATABASES")

dbGetQuery(spark, "SHOW TABLES")

dbGetQuery(spark, "DESCRIBE airlines")

dbGetQuery(spark, "SELECT * from airlines limit 10")


# `dbGetQuery()` returns the query result to R as a data frame

airlines <- dbGetQuery(spark, "SELECT * FROM airlines")

class(airlines)

airlines

# Only use `dbGetQuery()` when the query result will be small enough to fit in memory in your R session.


# You can also return the result of a SQL query as a `tbl_spark`.
# To do this, you need to load dplyr and use `tbl(spark, sql(...))`.
# This will *not* return the full query result; it will only display the first few rows.

library(dplyr)

flights <- tbl(spark, sql("SELECT * FROM flights"))

class(flights)

flights


# But you need not use SQL to access Hive tables with sparklyr
# Instead you can just reference the Hive table name with `tbl()`:

flights <- tbl(spark, "flights")

# This gives exactly the same result:

class(flights)

flights



# There are more details in the next module about how sparklyr works together with dplyr


# ## Copying data frames from R to Spark

# Use the `copy_to()` function to copy a local R data frame to Spark

iris_tbl <- copy_to(spark, iris)

# On our class cluster, this may fail when you are connected to a local Spark instance.
# Workaround: Connect to Spark on YARN instead.

spark_disconnect(spark)
spark <- spark_connect(master = "yarn", app_name = "write")
iris_tbl <- copy_to(spark, iris)

iris_tbl

# If you remove the variable `iris_tbl` (which represents this remote Spark data frame) you can re-create it without copying the data to Spark again.
# Just use `tbl()` and reference the name (`iris`):

rm(iris_tbl)

iris_tbl <- tbl(spark, "iris")

iris_tbl

# Note: `copy_to()` *does not persist* the copied data in HDFS.
# The data is stored in a temporary location in HDFS and may be cached in Spark memory.
# After you end your session by disconnecting from Spark, it will no longer be available.


# ## Cleanup

# Remove practice directory from HDFS:

system("hdfs dfs -rm -r -skipTrash practice/riders_tsv")
system("hdfs dfs -rm -r -skipTrash practive/riders_parquet")

# Stop the `SparkSession`:

spark_disconnect(spark)
